### PR TITLE
Stream Executor Changes for the Fused Conv+Bias+Actv implementation

### PR DESCRIPTION
**Pull request by deven-amd**

### CODE REVIEW

**Comment by deven-amd** [comment]
> @deven-amd would this work with stock MIOpen from ROCm 1.9 release? if not, could you help amend `Dockerfile.rocm`?

It will not...it needs a fix for a bug I reported yesterday. The fix is now in the `develop` branch of the private MLOpen repo, so we will need to pull in MIOpen from there.  Do you want the `Dockerfile.rocm` update as a separate PR or can I roll it into this one?

**Comment by whchung** [comment]
@deven-amd let's hear from @daniellowell . I'm not sure about the policy disclosing yet released MIOpen logic.

File: tensorflow/stream_executor/rocm/rocm_dnn.cc [hunk 1]
```diff
@@ -592,6 +730,279 @@ class ScopedNormalizeDescriptor {
...
+
+      case dnn::ActivationMode::kRelu6:
+	miopen_activation_mode = miopenActivationRELU;
+	alpha = 6.0;
+	break;
+
+      case dnn::ActivationMode::kTanh:
+	miopen_activation_mode = miopenActivationTANH;
+	break;
+
```
Comment (whchung) [inline tensorflow/stream_executor/rocm/rocm_dnn.cc]: what about kBandPass?

**Comment by deven-amd** [comment]
> @deven-amd since `dnn.h` is changed. probably need a dummy implementation on CUDA path if we want this be upstreamed?

there code the `dnn.h` has a `return false;` which should suffice as the default/dummy implementation I think.

File: tensorflow/stream_executor/rocm/rocm_dnn.cc (hunk 1 above)
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/rocm_dnn.cc]: there does not seem to be a TF OP for it, dont know how to access its `X` attribute (which specifies the `[-X,X]` range, and hence left it out

File: tensorflow/stream_executor/rocm/rocm_dnn.cc [hunk 2]
```diff
@@ -43,6 +43,8 @@ limitations under the License.
 #include "rocm/include/miopen/miopen.h"
 // clang-format on
 
+#include "tensorflow/core/lib/hash/hash.h"
```
Comment (whchung) [inline tensorflow/stream_executor/rocm/rocm_dnn.cc]: nit: probably need to group it with other tensorflow headers

File: tensorflow/stream_executor/rocm/rocm_dnn.cc [hunk 3]
```diff
@@ -195,6 +215,124 @@ MIOPEN_DNN_ROUTINE_EACH(PERFTOOLS_GPUTOOLS_MIOPEN_WRAP)
...
+    return hashValue;
+  }
+
+  // class to implement a cache of compiled fusion plans
+  class CachedFusionPlans {
+
+  public:
+
+    // check if we already have a fusion_plan corresponding to the given hash value
+    // If we fo
```
Comment (whchung) [inline tensorflow/stream_executor/rocm/rocm_dnn.cc]: fo?

File: tensorflow/stream_executor/rocm/rocm_dnn.cc [hunk 4]
```diff
@@ -195,6 +215,124 @@ MIOPEN_DNN_ROUTINE_EACH(PERFTOOLS_GPUTOOLS_MIOPEN_WRAP)
...
+    //   return false (+ newly created fusion plan via given pointer)
+    static bool FindOrCreate(uint64 hash, ROCMExecutor* parent, miopenFusionPlanDescriptor_t *fusion_plan,
+			     miopenFusionDirection_t fusion_direction, miopenTensorDescriptor_t input_descriptor) {
+
+      mutex_lock lock{cachedPlansMutex};
+
+      bool foundCachedPlan = false;
+
+      auto it = cachedPlans.find(hash);
+      if (it != cachedPlans.end()) {
```
Comment (whchung) [inline tensorflow/stream_executor/rocm/rocm_dnn.cc]: could you re-indent the code with `git clang-format`?

File: tensorflow/stream_executor/rocm/rocm_dnn.cc [hunk 5]
```diff
@@ -3507,6 +3918,106 @@ bool MIOpenSupport::DeriveOutputBatchDescriptor(
...
+    }
+    timer->Destroy();
+  }
+
+  if (status != miopenStatusSuccess) {
+    // Silently return when we are profiling.
+    if (!is_profiling) {
+      LOG(FATAL) << "failed to enqueue fused-convolution on stream: " << ToString(status);
+      return false;
+    }
```
Comment (whchung) [inline tensorflow/stream_executor/rocm/rocm_dnn.cc]: missing `else` path?

File: tensorflow/stream_executor/rocm/rocm_dnn.cc [hunk 6]
```diff
@@ -592,6 +730,279 @@ class ScopedNormalizeDescriptor {
...
+		 << ToString(status);
+    }
+
+    float alpha = 1.0;
+    float beta = 0.0;
+
+    return ScopedFusionPlanBase::SetActivationArgs(k_actv_op_idx, &alpha, &beta, activ_alpha, activ_beta, activ_gamma);
+  }
+
+  uint64 GetFusionOpHashValue(miopenTensorDescriptor_t input_descriptor,
```
Comment (whchung) [inline tensorflow/stream_executor/rocm/rocm_dnn.cc]: should this be a member function, or put inside an anonymous namespace just like others?

File: tensorflow/stream_executor/rocm/rocm_dnn.cc (hunk 3 above)
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/rocm_dnn.cc]: should be "do"...will fix and update PR

File: tensorflow/stream_executor/rocm/rocm_dnn.cc (hunk 2 above)
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/rocm_dnn.cc]: will be fixed in next commit

File: tensorflow/stream_executor/rocm/rocm_dnn.cc (hunk 4 above)
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/rocm_dnn.cc]: will be part of next commit...
just ran it...now I realize why all the TF code does not contain long lines...have to say not a fan of clang-format, makes code difficult too read!

File: tensorflow/stream_executor/rocm/rocm_dnn.cc (hunk 6 above)
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/rocm_dnn.cc]: this should be a member function, as each fusion op class should have one. almost made it an abstract virtual function, but then tht would required calling a virtual in a constructor, and that is rightfully frowned upon

File: tensorflow/stream_executor/rocm/rocm_dnn.cc (hunk 5 above)
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/rocm_dnn.cc]: no...return statement should be outside of `if (!is_profiling)`....good catch...thank you

**Comment by deven-amd** [comment]
@whchung 

now that `Dockerfile.rocm` has been updated, let me know if there is anything else that needs to be done prior to merging this PR.

thanks

**Comment by whchung** [comment]
I’ll spend some time today to give this PR another pass

**Pull request merged**